src/libs/type-utils/lodash-fp-helpers.test.ts

```diff
@@ -0,0 +1,451 @@
+      const handler2 = <F extends AnyPromiseFn, P>(
+        fn: GenericPromiseFn<F, P>
+      ): GenericPromiseFn<F, P> => {
+        return async (...args: Parameters<F>): Promise<P> => {
+          // simulate typical pattern of calling fn, then do handling
+          const result = await fn(...args)
+          handlersWatcher('handler2')
+          return result
+        }
+      }
+      const watcher = jest.fn()
+
+      const mainFn = async (a: string, b: string): Promise<string> => {
+        watcher(a, b)
+        await delay(100)
+        return `${a} ${b}!`
+      }
+
+      // Act
+      const myFn = withHandlers([handler1, handler2], mainFn)
+
+      const myResult = await myFn('hello', 'there')
+
+      // Assert
+      expect(myResult).toBe('hello there!')
+      expect(watcher).toBeCalledTimes(1)
+      expect(watcher).toBeCalledWith('hello', 'there')
+
+      expect(handlersWatcher).toBeCalledTimes(2)
+      expect(handlersWatcher).toHaveBeenNthCalledWith(1, 'handler1')
+      expect(handlersWatcher).toHaveBeenNthCalledWith(2, 'handler2')
+    })
+
+    it('handles handler wrapper promise-fns created with curryLastArg', async () => {
+      // Arrange
+      const handlersWatcher = jest.fn()
+
+      const handler1Fn = <F extends AnyPromiseFn, P>(
+        arg: string,
+        fn: GenericPromiseFn<F, P>
+      ): GenericPromiseFn<F, P> => {
+        return async (...args: Parameters<F>): Promise<P> => {
+          // simulate typical pattern of calling fn, then do handling
+          const result = await fn(...args)
+          handlersWatcher('handler1', arg)
+          return result
+        }
+      }
+
+      const handler1 = curryLastArg(handler1Fn)
+
+      const handler2Fn = <F extends AnyPromiseFn, P>(
+        arg1: string,
+        arg2: number,
+        fn: GenericPromiseFn<F, P>
+      ): GenericPromiseFn<F, P> => {
+        return async (...args: Parameters<F>): Promise<P> => {
+          // simulate typical pattern of calling fn, then do handling
+          const result = await fn(...args)
+          handlersWatcher('handler2', arg1, arg2)
+          return result
+        }
+      }
+
+      const handler2 = curryLastArg(handler2Fn)
+
+      const watcher = jest.fn()
+      const mainFn = async (a: string, b: string): Promise<string> => {
+        watcher(a, b)
+        await delay(100)
+        return `${a} ${b}!`
+      }
+
+      // Act
+      const myFn = withHandlers([
+        handler1('hi'),
+        handler2('sup', 7)
+      ],
+      mainFn)
+
+      const myResult = await myFn('hello', 'there')
+
+      // Assert
+      expect(myResult).toBe('hello there!')
+      expect(watcher).toBeCalledTimes(1)
+      expect(watcher).toBeCalledWith('hello', 'there')
+
+      expect(handlersWatcher).toBeCalledTimes(2)
+      expect(handlersWatcher).toHaveBeenNthCalledWith(1, 'handler1', 'hi')
+      expect(handlersWatcher).toHaveBeenNthCalledWith(2, 'handler2', 'sup', 7)
+    })
+
+    /* top recommendation */
```

Comment:
Should these tests be reordered to place this at the top of the file where it's most easily findable for reference?

Reply:
right now the ordering is basic, to flexible with safety, to safest (recommended), then SafeCurryX for completeness...  I could put SafeCurryX as first example to keep the progression.... the intent was to have the example show the progression and that they build on each other as options.

Reply:
That makes sense. Making this easier to find could also be accomplished with a link in the wiki.

Reply:
most IDE's also let you easily collapse code blocks... which I often do to collapse a set of tests one by one to see the listing more easily.